File: lib/test/t_SimplicialCubature_std.cxx
//                                               -*- C++ -*-
/**
 *  @brief The test file of class GaussKronrod
 *
 *  Copyright 2005-2026 Airbus-EDF-IMACS-ONERA-Phimeca
 *
 *  This library is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with this library.  If not, see <http://www.gnu.org/licenses/>.
 *
 */
#include "openturns/OT.hxx"
#include "openturns/OTtestcode.hxx"
#include <cmath>

using namespace OT;
using namespace OT::Test;

typedef Collection<Complex> ComplexCollection;

int main(int, char *[])
{
  TESTPREAMBLE;
  OStream fullprint(std::cout);
  PlatformInfo::SetNumericalPrecision(8);

  try
  {
    const Point ref = {0, 0, 1.0, 0.3591409, 0.09390606, 0.01935569};
    for (UnsignedInteger n = 2; n < 6; ++ n)
    {
      String sum = "x0";
      for (UnsignedInteger i = 1; i < n; ++ i)
        sum += (OSS() << "+x" << i);
      SymbolicFunction f(Description::BuildDefault(n, "x"), Description({"exp(" + sum + ")"}));
      fullprint << "f=" << f.__str__() << std::endl;
      Sample vertices(1, n);
      for (UnsignedInteger j = 0; j < n; ++ j)
      {
        Point e(n);
        e[j] = 1.0;
        vertices.add(e);
      }
      Indices indices(n + 1);
      indices.fill();
      IndicesCollection simplicies(Collection<Indices>(1, indices));
      Mesh canonicalSimplex(vertices, simplicies, true);
      for (UnsignedInteger k = 1; k < 5; ++ k)
      {
        SimplicialCubature algo;
        algo.setRule(k);
        // fullprint << "Algo=" << algo << std::endl;
        const Scalar value = algo.integrate(f, canonicalSimplex)[0];
        fullprint << "n=" << n << " k=" << k << " value=" << std::setprecision(16) << value << std::endl;
        assert_almost_equal(value, ref[n]);
      }
    }
    // The integration over a uniform mesh of a symmetric integrand produces
    // many simplices with equal errors, so the selection of the simplices to
    // refine must break the ties with the simplex index: the result has to be
    // identical on every call and accurate.
    {
      IntervalMesher mesher(Indices(3, 8));
      Mesh mesh(mesher.build(Interval(Point(3, 0.0), Point(3, 1.0))));
      SymbolicFunction f(Description::BuildDefault(3, "x"), Description({"exp(-10.0*(x0+x1+x2))"}));
      const Scalar refValue = std::pow((1.0 - std::exp(-10.0)) / 10.0, 3);
      SimplicialCubature algo;
      algo.setRule(3);
      algo.setMaximumRelativeError(1.0e-7);
      const Point reference(algo.integrate(f, mesh));
      const UnsignedInteger callsAfterFirst = f.getCallsNumber();
      for (UnsignedInteger i = 0; i < 3; ++ i)
      {
        const Point value(algo.integrate(f, mesh));
        assert_almost_equal(value, reference);
        assert_almost_equal(value[0], refValue, 1.0e-6, 1.0e-8);
      }
      // The refinement must take place for the tie-breaking to be exercised
      const UnsignedInteger callsPerIntegration = f.getCallsNumber() - callsAfterFirst;
      if (!(callsPerIntegration > mesh.getSimplicesNumber() * 16))
        throw TestFailed(OSS() << "No refinement occurred, calls=" << callsPerIntegration);
      fullprint << "tie-breaking calls=" << f.getCallsNumber() << " value=" << std::setprecision(16) << reference[0] << std::endl;
    }
    // The batched evaluation must give the same result for any block size
    {
      IntervalMesher mesher(Indices(3, 8));
      Mesh mesh(mesher.build(Interval(Point(3, 0.0), Point(3, 1.0))));
      SymbolicFunction f(Description::BuildDefault(3, "x"), Description({"exp(-10.0*(x0+x1+x2))"}));
      const Scalar refValue = std::pow((1.0 - std::exp(-10.0)) / 10.0, 3);
      SimplicialCubature algo;
      algo.setRule(3);
      algo.setMaximumRelativeError(1.0e-6);
      Point reference;
      const UnsignedInteger savedBlockSize = ResourceMap::GetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize");
      try
      {
        for (const UnsignedInteger blockSize : {2048u, 7u, 1u})
        {
          ResourceMap::SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", blockSize);
          const Point value(algo.integrate(f, mesh));
          if (reference.getDimension() == 0) reference = value;
          assert_almost_equal(value[0], refValue, 1.0e-6, 1.0e-8);
          assert_almost_equal(value, reference, 1.0e-12, 0.0);
        }
      }
      catch (...)
      {
        ResourceMap::SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", savedBlockSize);
        throw;
      }
      ResourceMap::SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", savedBlockSize);
    }
    // The invalid inputs must be rejected
    {
      SimplicialCubature algo;
      Sample vertices(1, 2);
      vertices.add(Point({1.0, 0.0}));
      vertices.add(Point({0.0, 1.0}));
      Indices indices(3);
      indices[0] = 0;
      indices[1] = 1;
      indices[2] = 2;
      Mesh mesh(vertices, IndicesCollection(Collection<Indices>(1, indices)));
      bool caught = false;
      try
      {
        algo.integrate(SymbolicFunction(Description(1, "x"), Description(1, "x")), mesh);
      }
      catch (const InvalidArgumentException &)
      {
        caught = true;
      }
      assert_almost_equal(caught ? 1.0 : 0.0, 1.0);
      // An empty mesh of the right dimension integrates to zero
      Mesh emptyMesh(Sample(0, 2), IndicesCollection(0, 0));
      const Point value(algo.integrate(SymbolicFunction(Description::BuildDefault(2, "x"), Description({"x0", "x1"})), emptyMesh));
      assert_almost_equal(value, Point(2, 0.0));
      // A null block size must be rejected
      {
        const UnsignedInteger savedBlockSize = ResourceMap::GetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize");
        caught = false;
        try
        {
          ResourceMap::SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", 0);
          algo.integrate(SymbolicFunction(Description::BuildDefault(2, "x"), Description(1, "x0")), mesh);
        }
        catch (const InvalidArgumentException &)
        {
          caught = true;
        }
        catch (...)
        {
          ResourceMap::SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", savedBlockSize);
          throw;
        }
        ResourceMap::SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", savedBlockSize);
      }
      assert_almost_equal(caught ? 1.0 : 0.0, 1.0);
    }
  }
  catch (TestFailed & ex)
  {
    std::cerr << ex << std::endl;
    return ExitCode::Error;
  }
  return ExitCode::Success;
}

File: python/test/t_SimplicialCubature_std.py
#! /usr/bin/env python

import openturns as ot
import openturns.testing as ott
import math as m

v1 = [0.0, 0.0, 0.0]
v2 = [1.0, 1.0, 1.0]
v3 = [0.0, 1.0, 1.0]
v4 = [0.0, 0.0, 1.0]
# simplex
S = [v1, v2, v3, v4]
simplicies = [[0, 1, 2, 3]]
mesh = ot.Mesh(S, simplicies)
# function to integrate
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(x1 + x2 + x3)"])
# integral of f
algo = ot.SimplicialCubature()
value = algo.integrate(f, mesh)[0]
print(value)
ott.assert_almost_equal(value, (m.exp(1.0) - 1.0) ** 3 / 6)

# Test with interval inferface
f = ot.SymbolicFunction(["x", "y", "z"], ["sin(x) * cos(y) * exp(z)"])
valueRef = -m.sin(1.0) * (m.cos(1.0) - 1.0) * (m.e - 1.0)
value = algo.integrate(f, ot.Interval([0.0] * 3, [1.0] * 3))
ott.assert_almost_equal(value[0], valueRef)

# Multi-output integrand over the simplex
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["x1", "x2"])
value = algo.integrate(f, mesh)
ott.assert_almost_equal(value, [1.0 / 24.0, 1.0 / 12.0])

# Adaptive refinement on a peaked integrand over [0, 1]^3
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(-10.0 * (x1 + x2 + x3))"])
valueRef = ((1.0 - m.exp(-10.0)) / 10.0) ** 3
algo.setRule(3)
algo.setMaximumRelativeError(1e-5)
value = algo.integrate(f, ot.Interval([0.0] * 3, [1.0] * 3))
ott.assert_almost_equal(value[0], valueRef, 1e-6, 1e-12)

# Rule 4 accuracy over the simplex
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(x1 + x2 + x3)"])
algo.setRule(4)
algo.setMaximumRelativeError(1e-5)
value = algo.integrate(f, mesh)[0]
ott.assert_almost_equal(value, (m.exp(1.0) - 1.0) ** 3 / 6, 1e-6, 1e-12)

# Rules 1 and 2 accuracy over the simplex and over a 2D interval
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(x1 + x2 + x3)"])
refSimplex = (m.exp(1.0) - 1.0) ** 3 / 6
f2 = ot.SymbolicFunction(["x", "y"], ["sin(x) * cos(y)"])
refInterval2 = -m.sin(1.0) * (m.cos(1.0) - 1.0)
for rule in [1, 2]:
    algo.setRule(rule)
    algo.setMaximumRelativeError(1e-5)
    value = algo.integrate(f, mesh)[0]
    ott.assert_almost_equal(value, refSimplex, 1e-6, 1e-12)
    value = algo.integrate(f2, ot.Interval([0.0] * 2, [1.0] * 2))[0]
    ott.assert_almost_equal(value, refInterval2, 1e-6, 1e-12)

# Adaptive refinement of a multi-output integrand: each marginal is integrated
# independently with its own error estimates
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(-10.0 * x1)", "exp(-10.0 * x2)", "exp(-10.0 * x3)"])
valueRef = ((1.0 - m.exp(-10.0)) / 10.0) * ot.Point([1.0, 1.0, 1.0])
algo.setRule(3)
algo.setMaximumRelativeError(1e-5)
value = algo.integrate(f, ot.Interval([0.0] * 3, [1.0] * 3))
ott.assert_almost_equal(value, valueRef, 1e-6, 1e-12)

# Regression test for the deterministic selection of the simplices to refine:
# a uniform mesh of a symmetric integrand produces many simplices with equal
# errors, so the partial sort must break the ties with the simplex index to
# select a deterministic set. The result must be identical on every call and
# accurate.
meshUniform = ot.IntervalMesher([8, 8, 8]).build(ot.Interval([0.0] * 3, [1.0] * 3))
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(-10.0 * (x1 + x2 + x3))"])
valueRef = ((1.0 - m.exp(-10.0)) / 10.0) ** 3
algo.setRule(3)
algo.setMaximumRelativeError(1e-7)
valueRefAdap = algo.integrate(f, meshUniform)
for _ in range(3):
    value = algo.integrate(f, meshUniform)
    ott.assert_almost_equal(value, valueRefAdap)
    ott.assert_almost_equal(value[0], valueRef, 1e-6, 1e-8)
# The refinement must actually take place for the tie-breaking to be exercised
assert f.getCallsNumber() > meshUniform.getSimplicesNumber() * 16

# The batched evaluation must give the same result for any block size
savedBlockSize = ot.ResourceMap.GetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize")
try:
    f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(-10.0 * (x1 + x2 + x3))"])
    algo.setMaximumRelativeError(1e-6)
    reference = None
    for blockSize in [2048, 7, 1]:
        ot.ResourceMap.SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", blockSize)
        value = algo.integrate(f, meshUniform)
        ott.assert_almost_equal(value[0], valueRef, 1e-6, 1e-8)
        if reference is None:
            reference = value
        ott.assert_almost_equal(value, reference, 0.0, 1e-12)
finally:
    ot.ResourceMap.SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", savedBlockSize)

# Invalid inputs are rejected
ot.ResourceMap.SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", 2048)
with ott.assert_raises(TypeError):
    algo.integrate(ot.SymbolicFunction(["x"], ["x"]), mesh)
with ott.assert_raises(TypeError):
    algo.setRule(5)
with ott.assert_raises(TypeError):
    ot.ResourceMap.SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", 0)
    algo.integrate(f, meshUniform)
ot.ResourceMap.SetAsUnsignedInteger("SimplicialCubature-EvaluationBlockSize", 2048)

# An empty mesh of the right dimension integrates to zero
f = ot.SymbolicFunction(["x1", "x2"], ["x1", "x2"])
emptyMesh = ot.Mesh(ot.Sample(0, 2), ot.IndicesCollection([]))
value = algo.integrate(f, emptyMesh)
ott.assert_almost_equal(value, [0.0, 0.0])

# The maximum calls number bounds the number of evaluations
f = ot.SymbolicFunction(["x1", "x2", "x3"], ["exp(-20.0 * (x1 + x2 + x3))"])
algo.setRule(3)
algo.setMaximumCallsNumber(200)
algo.integrate(f, ot.Interval([0.0] * 3, [1.0] * 3))
assert f.getCallsNumber() >= 1
assert f.getCallsNumber() <= 4 * 200

# Accessor round-trip tests
algo2 = ot.SimplicialCubature()
algo2.setRule(2)
assert algo2.getRule() == 2
algo2.setMaximumAbsoluteError(1e-10)
ott.assert_almost_equal(algo2.getMaximumAbsoluteError(), 1e-10)
algo2.setMaximumRelativeError(1e-8)
ott.assert_almost_equal(algo2.getMaximumRelativeError(), 1e-8)
algo2.setMaximumCallsNumber(500)
assert algo2.getMaximumCallsNumber() == 500

# repr/str do not crash
r = repr(algo2)
s = str(algo2)
assert "SimplicialCubature" in s
assert "rule" in s

# Default constructor uses ResourceMap defaults
algo3 = ot.SimplicialCubature()
assert algo3.getRule() == ot.ResourceMap.GetAsUnsignedInteger("SimplicialCubature-DefaultRule")
ott.assert_almost_equal(algo3.getMaximumAbsoluteError(), ot.ResourceMap.GetAsScalar("SimplicialCubature-DefaultMaximumAbsoluteError"))
ott.assert_almost_equal(algo3.getMaximumRelativeError(), ot.ResourceMap.GetAsScalar("SimplicialCubature-DefaultMaximumRelativeError"))
assert algo3.getMaximumCallsNumber() == ot.ResourceMap.GetAsUnsignedInteger("SimplicialCubature-DefaultMaximumCallsNumber")

# 1D integration over an interval
f1d = ot.SymbolicFunction(["x"], ["x^2"])
value1d = algo2.integrate(f1d, ot.Interval([0.0], [1.0]))
ott.assert_almost_equal(value1d[0], 1.0 / 3.0)

# 2D integration over a triangle
f2d = ot.SymbolicFunction(["x", "y"], ["x + y"])
v1 = [0.0, 0.0]
v2 = [1.0, 0.0]
v3 = [0.0, 1.0]
mesh2d = ot.Mesh([v1, v2, v3], [[0, 1, 2]])
value2d = algo2.integrate(f2d, mesh2d)
ott.assert_almost_equal(value2d[0], 1.0 / 3.0)
